src/IO/S3/Client.cpp

```diff
@@ -741,88 +742,123 @@ Client::doRequest(RequestType & request, RequestFn request_fn) const
+                /*name*/ "",
+                /*message*/ fmt::format("All {} retry attempts failed. Last exception: {}", max_attempts, getCurrentExceptionMessage(false)),
+                /*retryable*/ true);
+
+            // network exceptions are always retryable, we could just return true here
+            // but we have to checks cancelation points for query
+            return !CurrentThread::isInitialized() || !CurrentThread::get().isQueryCanceled();
+        };
+
+        for (attempt_no = 0; attempt_no < max_attempts; ++attempt_no)
         {
             incrementProfileEvents<IsReadMethod>(ProfileEvents::S3ReadRequestAttempts, ProfileEvents::S3WriteRequestAttempts);
             if (isClientForDisk())
                 incrementProfileEvents<IsReadMethod>(ProfileEvents::DiskS3ReadRequestAttempts, ProfileEvents::DiskS3WriteRequestAttempts);
 
+            if (attempt_no > 0)
+            {
+                incrementProfileEvents<IsReadMethod>(ProfileEvents::S3ReadRequestRetryableErrors, ProfileEvents::S3WriteRequestRetryableErrors);
+                if (isClientForDisk())
+                    incrementProfileEvents<IsReadMethod>(ProfileEvents::DiskS3ReadRequestRetryableErrors, ProfileEvents::DiskS3WriteRequestRetryableErrors);
+
+                // use previusly attempt number to calculate delay
+                updateNextTimeToRetryAfterRetryableError(outcome.GetError(), attempt_no - 1);
+
+                // update ClickHouse-specific attempt number in the request
+                // to help choose the right timeouts on the HTTP client which depends on retry attempt number
+                auto clickhouse_request_attempt = getClickhouseAttemptNumber(request_);
+                setClickhouseAttemptNumber(request_, clickhouse_request_attempt + attempt_no);
+            }
+
             /// Slowing down due to a previously encountered retryable error, possibly from another thread.
             slowDownAfterRetryableError();
 
             try
             {
-                /// S3 does retries network errors actually.
-                /// But it does matter when errors occur.
-                /// This code retries a specific case when
-                /// network error happens when XML document is being read from the response body.
-                /// Hence, the response body is a stream, network errors are possible at reading.
-                /// S3 doesn't retry them.
-
-                /// Not all requests can be retried in that way.
-                /// Requests that read out response body to build the result are possible to retry.
-                /// Requests that expose the response stream as an answer are not retried with that code. E.g. GetObject.
-                auto outcome = request_fn_(request_);
-
-                if (!outcome.IsSuccess()
-                    /// AWS SDK's built-in per-thread retry logic is disabled.
-                    && client_configuration.s3_slow_all_threads_after_retryable_error
```

Comment:
If seems like when `s3_slow_all_threads_after_retryable_error` is `false, we do not do any retries at all here.